### PR TITLE
CXL DVSEC Capabilities Decoded

### PR DESCRIPTION
**Pull request by jphaws**
This series decodes the remainder of the CXL capabilities that were left undecoded in #110. This includes:

- GPF DVSEC for CXL Devices (Section 8.1.7 of the CXL 3.0 spec)
- PCIe DVSEC for Flex Bus Port (Section 8.1.8 of the CXL 3.0 spec)
- GPF DVSEC for CXL Ports (Section 8.1.6 of the CXL 3.0 spec)
- MLD DVSEC (Section 8.1.10 of the CXL 3.0 spec)
- Non-CXL Function Map DVSEC (Section 8.1.4 of the CXL 3.0 spec)

Also includes a configuration space dump added to `tests/` of the CXL device I have access to.

### CODE REVIEW

**Review by gollux — CHANGES_REQUESTED** [review]
Generally, I like this, but I left a couple of nit-picking comments.

File: ls-ecaps.c [hunk 1]
```diff
@@ -880,6 +880,67 @@ dvsec_cxl_gpf_device(struct device *d, int where)
...
+  printf("\n\t\tFBModTS:\tReceived FB Data: %06x\n", (unsigned int)data);
+
+  if (rev > 1){
+    u8 nop;
+
+    l = get_conf_long(d, where + PCI_CXL_FB_PORT_CAP2);
+    printf("\t\tFBCap2:\t NOPHint%c\n", FLAG(l, PCI_CXL_FB_CAP2_NOP_HINT));
+
+    l = get_conf_long(d, where + PCI_CXL_FB_PORT_CTRL2);
+    printf("\t\tFBCap2:\t NOPHint%c\n", FLAG(l, PCI_CXL_FB_CTRL2_NOP_HINT));
```
[the remark this answers is not in the document]
Comment (jphaws) [reply ls-ecaps.c]: No you are correct, that should have been `FBCtl2` and `FBSta2` below. Fixed that as well as space after `\t ` for each as well.

File: ls-ecaps.c [hunk 2]
```diff
@@ -880,6 +880,67 @@ dvsec_cxl_gpf_device(struct device *d, int where)
   printf("\t\tGPF Phase 2 Power: %umW\n", (unsigned int)l);
 }
 
+static void
+dvsec_cxl_flex_bus(struct device *d, int where, int rev)
+{
+  u16 w;
+  u32 l, data;
+
+  if (rev < 1)
```
[the remark this answers is not in the document]
Comment (jphaws) [reply ls-ecaps.c:958]: Added a warning: `Revision %d not supported`. Let me know if you'd like to see something else as the warning message or if that is acceptable.

File: ls-ecaps.c [hunk 3]
```diff
@@ -1022,6 +1022,36 @@ dvsec_cxl_mld(struct device *d, int where)
     printf("\t\tNumLogDevs: %d\n", w);
 }
 
+static void
+dvsec_cxl_function_map(struct device *d, int where)
+{
+  u32 l;
+
+  l = get_conf_word(d, where + PCI_CXL_FUN_MAP_REG_0);
```
[the remark this answers is not in the document]
Comment (jphaws) [reply ls-ecaps.c]: Fixed

**Comment by gollux** [comment]
Thanks, merged.

**Pull request merged**